components/brave_wallet/browser/simulation_service.cc

```diff
@@ -0,0 +1,326 @@
+       brave_wallet::kBlowfishAPIVersion},
+      {net::HttpRequestHeaders::kContentType, "application/json"},
+      {brave_wallet::kBraveServicesKeyHeader, BUILDFLAG(BRAVE_SERVICES_KEY)}};
+}
+
+}  // namespace
+
+namespace brave_wallet {
+
+GURL SimulationService::base_url_for_test_;
+
+SimulationService::SimulationService(
+    scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory,
+    JsonRpcService* json_rpc_service)
+    : api_request_helper_(GetNetworkTrafficAnnotationTag(), url_loader_factory),
+      json_rpc_service_(json_rpc_service),
+      weak_ptr_factory_(this) {
+  DCHECK(json_rpc_service_);
+}
+
+SimulationService::~SimulationService() = default;
+
+mojo::PendingRemote<mojom::SimulationService> SimulationService::MakeRemote() {
+  mojo::PendingRemote<mojom::SimulationService> remote;
+  receivers_.Add(this, remote.InitWithNewPipeAndPassReceiver());
+  return remote;
+}
+
+void SimulationService::Bind(
+    mojo::PendingReceiver<mojom::SimulationService> receiver) {
+  receivers_.Add(this, std::move(receiver));
+}
+
+void SimulationService::SetBaseURLForTest(const GURL& base_url_for_test) {
+  base_url_for_test_ = base_url_for_test;
+}
+
+// static
+GURL SimulationService::GetScanTransactionURL(const std::string& chain_id,
+                                              mojom::CoinType coin) {
+  std::string spec = base::StringPrintf(
+      "%s/%s/%s",
+      base_url_for_test_.is_empty() ? brave_wallet::kBlowfishBaseAPIURL
+                                    : base_url_for_test_.spec().c_str(),
+      GetRelativeScanURL(chain_id, coin).c_str(),
+      coin == mojom::CoinType::SOL ? "transactions" : "transaction");
+  GURL url(spec);
+  return url;
+}
+
+// static
+GURL SimulationService::GetScanMessageURL(const std::string& chain_id,
+                                          mojom::CoinType coin) {
+  std::string spec = base::StringPrintf(
+      "%s/%s/message",
+      base_url_for_test_.is_empty() ? brave_wallet::kBlowfishBaseAPIURL
+                                    : base_url_for_test_.spec().c_str(),
+      GetRelativeScanURL(chain_id, coin).c_str());
+  GURL url(spec);
+  return url;
+}
+
+void SimulationService::HasTransactionScanSupport(
+    const std::string& chain_id,
+    mojom::CoinType coin,
+    HasTransactionScanSupportCallback callback) {
+  bool result = HasTransactionScanSupportInternal(chain_id, coin);
+  std::move(callback).Run(result);
+}
+
+void SimulationService::HasMessageScanSupport(
+    const std::string& chain_id,
+    mojom::CoinType coin,
+    HasMessageScanSupportCallback callback) {
+  bool result = HasMessageScanSupportInternal(chain_id, coin);
+  std::move(callback).Run(result);
+}
+
+void SimulationService::ScanSolanaTransaction(
+    mojom::SolanaTransactionRequestUnionPtr request,
+    const std::string& language,
+    ScanSolanaTransactionCallback callback) {
+  const auto& chain_id =
+      json_rpc_service_->GetChainIdSync(mojom::CoinType::SOL, absl::nullopt);
```

Comment:
This is wrong, you have to use `chainId` from request, otherwise what you are getting here is only for global default not the selected network per origin one. In other words, this only works correct on request initiated from internal page like brave://wallet but not regular dapps that have valid origin info.

Reply:
Fixed in 012cb657ca1aa9a22c14fe0497d8e514f5a8e3eb. I was also able to remove dependency on `JsonRpcService`.